util/alertifier.go

```diff
@@ -0,0 +1,138 @@
+package util
+
+// DCSO FEVER
+// Copyright (c) 2020, DCSO GmbH
+
+import (
+	"fmt"
+
+	"github.com/DCSO/fever/types"
+
+	"github.com/buger/jsonparser"
+	log "github.com/sirupsen/logrus"
+)
+
+// ExtraModifier is a function type that describes a function that adds the
+// appropriate `_extra` sub-object entries to a EVE-JSON event.
+type ExtraModifier func(inputAlert *types.Entry, ioc string) error
+
+// AlertJSONProvider is an interface describing a component that returns an
+// `alert` JSON sub-object to use in an EVE-JSON event.
+type AlertJSONProvider interface {
+	// GetAlertJSON is a function that returns a byte slice containing the
+	// JSON data for an `alert` EVE-JSON sub-object.
+	GetAlertJSON(inputEvent types.Entry, prefix string, ioc string) ([]byte, error)
+}
+
+// Alertifier is a component that creates EVE-JSON alerts from arbitrary
+// EVE-JSON events. It does this by cloning the original event and adding
+// alert-specific fields, depending on the given ExtraModifier and a set of
+// AlertJSONProviders, selectable using a string tag.
+type Alertifier struct {
+	alertPrefix   string
+	extraModifier ExtraModifier
+	matchTypes    map[string]AlertJSONProvider
+}
+
+// MakeAlertifier returns a new Alertifier, with no AlertJSONProviders set for
+// any match types, but with the given alert prefix preconfigured.
+// The alert prefix is a string that is prepended to all alert.signature values,
+// as in "DCSO TIE-BLF" or "ETPRO CURRENT_EVENTS", etc.
+func MakeAlertifier(prefix string) *Alertifier {
+	a := &Alertifier{
+		alertPrefix: prefix,
+		matchTypes:  make(map[string]AlertJSONProvider),
+	}
+	return a
+}
+
+// RegisterMatchType associates a given AlertJSONProvider with a match type tag.
+// It makes it callable in the MakeAlert() function in this Alertifier.
+func (a *Alertifier) RegisterMatchType(matchTypeName string, mt AlertJSONProvider) {
+	a.matchTypes[matchTypeName] = mt
+}
+
+// SetPrefix sets the signature prefix of the current Alertifier to the given
+// string value.
+func (a *Alertifier) SetPrefix(prefix string) {
+	a.alertPrefix = prefix
+}
+
+// SetExtraModifier sets the _extra modifier of the current Alertifier to the
+// passed function. Set it to nil to disable modification of the _extra
+// sub-object.
+func (a *Alertifier) SetExtraModifier(em ExtraModifier) {
+	a.extraModifier = em
+}
+
+// MakeAlert generates a new Entry representing an `alert` event based on the
+// given input metadata event. It uses the information from the Alertifier as
+// well as the given IoC to craft an `alert` sub-object in the resulting
+// alert, which is built by the AlertJSONProvider registered under the specified
+// matchType.
+func (a *Alertifier) MakeAlert(inputEvent types.Entry, ioc string,
+	matchType string) (*types.Entry, error) {
+	if v, ok := a.matchTypes[matchType]; ok {
```

Comment:
IMHO, early exit makes this more readable:

```
v, ok := a.matchTyeps[matchType]
if !ok {
  return nil, fmt.Errorf("cannot create alert for metadata, unknown "+
		"matchtype '%s'", matchType)
}

// clone the original event
newEntry...
```

Reply:
True. Will change this before merge.